### ISSUE TITLE
R95.2 Safari on iPad 12.9 - Cloud save to Google

### ISSUE THREAD

**Issue by zumwaltwood** [Scirra/Construct-3-bugs#1506]
## Problem description

*Enter a concise description of your problem here*
Select time way to long before folder tree shows up
Use a default folder name instead like you do with DropBox such as Apps/Construct 3 and save/load from there, within that folder, make folders for the project(s) by giving the users options for a project name. For example, with Ghost Shooter example for cloud save to Google Drive (or anothers) have it save to Apps/Construct 3/Ghost Shooter, etc.

Google drive is not optimized for folder browsing and never remembers your last location you were in, not your fault, just a flaw in how google drive is organized.

## Attach a .c3p

*A minimal Construct 3 project (.c3p) is required to be attached. Please see the guidelines*

## Steps to reproduce

1.
2.
3.

## Observed result

*What do you see happen?*

## Expected result

*What did you expect to happen instead?*

## Affected browsers

* Chrome: yes/no
* Firefox: yes/no
* Edge: yes/no
* Safari: yes/no

## System details

*If you see a crash report dialog, please copy and paste it to where it says "PASTE HERE" below. Otherwise please go to **Menu** -> **About** -> **Platform information** and paste that information there instead.*

<details><summary>View details</summary>

PASTE HERE

</details>

**Comment by shortercode** [comment]
A few people have brought this up before, and I've done several upgrades to improve performance in this area.

Dropbox and Google Drive behave differently because of the configuration options we chose to use with them. Both services offer a **general** and **app** mode for accessing files. The difference being that for **app** your product is assigned a special folder and it can't read or write anything outside of that folder. I considered **app** mode generally preferable as some users may feel uncomfortable giving access to all their files. Unfortunately on Google Drive users are not given any access to the contents of their application folders, which makes all sorts of things difficult or impossible for a user to do. Hence we ended up using **app** mode on Dropbox and **general** mode on Google Drive.

Given the way the limitations work moving from **general** to **app** mode would make it nearly impossible for existing users to access their existing projects in construct 3, and would reintroduce limitations we didn't want users to have. So we won't be changing that.

In terms of using a "default" folder like you suggest, while it would possibly reduce a couple of mouse clicks for users it currently wouldn't reduce the loading time at all. Provided a user hasn't disabled caching on C3 it will only load the file tree when the dialog is opened the first time that session, but it will load the whole file tree when it does ( even if it only needs to show part of it ).

I've planned an upgrade at some point which would allow it to load parts of the file tree as required. That should in theory make it faster, but would mean that users would see the loading wheel a lot more.

**Comment by zumwaltwood** [comment]
The problem I run into is literally the file structure for google drive, I have personally been using google drive since it was invented and made publicly available, as such I have tons of data in it, when I finally found a way to use it with Scirra C3, I literally waited almost 2 minutes for it to show the tree and I was waiting on a timeout message.  

Possible solution, generate the folder and get the folderId, don't bother with getting the folder tree, set the driveService.files().create to always create in that folder, now for the sub folders, you have the parent folderId, so you can generate folder(s) for the projects and store the folderId with the project, the only problem this then poses, is that you have to store this information in the local browser somewhere and if they purge their cache then you have another load time to deal with.

Lastly, we can grab the folderId ourself and put it on the project since it is part of the drive.google.com/drive/folders/ ID

Just some thoughts, thinking this outloud looking at the API now.

**Comment by AshleyScirra** [comment]
@shortercode - is this related to #1257? (i.e. the requests take a long time when there's a lot of content on the account)

**Comment by shortercode** [comment]
@AshleyScirra it is yes. When we load the cloud dialog we request all the folders and C3P projects on the users cloud account and build a folder tree from the results. This is then kept for the entire session, and patched whenever we do an cloud operation. If people have **a lot** of content though Google Drive takes a long time to generate the reply.

In that bug I increased the request size from 100 to 1000 and made changes so that we don't do any pre-flight requests, which considerably reduced the loading time.

My next planned improvement is to load folders into the cache as needed, that way users won't have to download information on folders unrelated to C3. I would have done it already but it breaks some other features ( like periodic back ups ) and requires some restructuring.

@zumwaltwood changes like you suggest would cause a lot of trouble for existing users, and caching the entry GUIDs locally causes confusing synchronisation issues for users who move projects around outside of C3.

**Comment by shortercode** [comment]
Okay I've just finished a fairly major revision to the cloud save which should improve performance for you, which should be included in the next release. When you attempt to open the dialog now it will only ask for a list of entries in the root folder. Given the number of changes that this revision includes I would expect a few bugs but hopefully most of them will be fairly minimal.

As this added in a lot of loading screens while moving around folders I added in a pre-caching mechanism. When you view a folder, it will attempt to load entries for the next level down into the cache in the background. In practice on a fast connection the first few folders are loaded before I try to click on them, so no loading screen is shown.
